tasks.py

```diff
@@ -856,3 +925,300 @@ def tests(context, lint_only=False, keepdb=False):
+
+
+def _print_tests_results(path: Path) -> dict:
+    summary = {
+        "total_tests": 0,
+        "total_errors": 0,
+        "total_failures": 0,
+        "total_time": 0.0,
+    }
+
+    for file in path.glob("*.xml"):
+        try:
+            tree = ElementTree.parse(file)
+        except ElementTree.ParseError:
+            print(50 * "=")
+            print(f"ERROR: Failed to parse {file}")
+            print(50 * "=")
+            continue
+        root = tree.getroot()
+        summary["total_tests"] += int(root.attrib.get("tests", 0))
+        summary["total_errors"] += int(root.attrib.get("errors", 0))
+        summary["total_failures"] += int(root.attrib.get("failures", 0))
+        summary["total_time"] += float(root.attrib.get("time", 0.0))
+
+    print("Test Results Summary:")
+    print("---------------------")
+    print(f"Total tests:\t{summary['total_tests']}")
+    print(f"Total errors:\t{summary['total_errors']}")
+    print(f"Total failures:\t{summary['total_failures']}")
+    print(f"Total time (s):\t{round(summary['total_time'])}")
+
+    return summary
+
+
+def _distribute_tests(distribution: DistributionType, workers) -> Tuple[List[List[str]], List[float]]:
+    """Split tests into groups for parallelization based on stored times.
+
+    TIMES distribution is better, however tests are failing sometimes, so we are not using it for now.
+    """
+
+    nautobot_path = _ROOT_PATH / "nautobot"
+    root_dir = f"{str(_ROOT_PATH)}/"
+
+    # Read all tests files removing the ones we want to skip
+    all_tests_files = [
+        str(path).replace(root_dir, "")
+        for path in nautobot_path.rglob("test_*.py")
+        if all(skip_name not in str(path) for skip_name in _SKIP_TESTS_NAMES)
+    ]
+
+    # Load stored tests times
+    tests_times = json.loads(_TESTS_TIMES_PATH.read_text(encoding="utf-8"))
+
+    # For NAMES distribution, group count is specified by the number of names in the distribution
+    group_count = len(_NAMES_DISTRIBUTION) if distribution == DistributionType.NAMES else workers
+
+    groups = [[] for _ in range(group_count)]
+    estimations = [0.0 for _ in range(group_count)]
+
+    obsoleted = []
+
+    def add_file(filename: str, estimation: float, group_index=None) -> None:
+        if group_index is None:
+            group_index = estimations.index(min(estimations))
+        try:
+            all_tests_files.remove(filename)
+        except ValueError:
+            print(f"WARNING: {filename} was not found in all_tests_files, skipping.")
+            obsoleted.append(filename)
+            return
+        groups[group_index].append(filename[:-3].replace("/", "."))
+        estimations[group_index] += estimation
+
+    if distribution == DistributionType.NAMES:
+        # Sort more specific items first
+        names = sorted(_NAMES_DISTRIBUTION, reverse=True)
+        tests_times_len = len(tests_times)
+        for index, name in enumerate(names):
+            for item in tests_times[:]:
```

Comment:
I'm not familiar with the `[:]` notation, what does it do?

Reply:
It's a copy of the list, used because the algorithm is removing items from the original one. What's your preferred notation? Would `[*original_list]` be better readable?